compiler/rustc_trait_selection/src/solve/project_goals.rs

```diff
@@ -308,13 +308,22 @@ impl<'tcx> assembly::GoalKind<'tcx> for ProjectionPredicate<'tcx> {
     ) -> QueryResult<'tcx> {
         let tcx = ecx.tcx();
         let Some(tupled_inputs_and_output) =
-        structural_traits::extract_tupled_inputs_and_output_from_callable(
-            tcx,
-            goal.predicate.self_ty(),
-            goal_kind,
-        )? else {
-        return ecx.make_canonical_response(Certainty::AMBIGUOUS);
-    };
+            structural_traits::extract_tupled_inputs_and_output_from_callable(
+                tcx,
+                goal.predicate.self_ty(),
+                goal_kind,
+            )? else {
+            return ecx.make_canonical_response(Certainty::AMBIGUOUS);
+        };
+
+        // FIXME(non_lifetime_binders): Higher-ranked Fn trait candidates are not (yet) supported.
+        // Make sure that the inputs/output don't capture any placeholder types.
```

Comment:
I disagree with this. Canonicalization also maps `ty::Param` to a placeholder.

I generally think that this check should happen somewhere else :thinking: